src/server/editorServices.ts

```diff
@@ -358,6 +358,67 @@ namespace ts.server {
+                getDirectories: fshost.getDirectories.bind(fshost),
+                getCurrentDirectory: fshost.getCurrentDirectory.bind(fshost),
+                getExecutingFilePath: fshost.getExecutingFilePath.bind(fshost),
+                realpath: fshost.realpath?.bind(fshost),
+                resolvePath: fshost.resolvePath.bind(fshost),
+                createDirectory: fshost.createDirectory.bind(fshost),
+                setModifiedTime: fshost.setModifiedTime?.bind(fshost),
+                deleteFile: fshost.deleteFile?.bind(fshost),
+                readDirectory: fshost.readDirectory.bind(fshost),
+                watchFile: fshost.watchFile.bind(fshost),
+                watchDirectory: fshost.watchDirectory.bind(fshost),
+                useCaseSensitiveFileNames: fshost.useCaseSensitiveFileNames,
+                // ServerHost
+                setTimeout: host.setTimeout.bind(host),
+                clearTimeout: host.clearTimeout.bind(host),
+                setImmediate: host.setImmediate.bind(host),
+                clearImmediate: host.clearImmediate.bind(host),
+                gc: host.gc?.bind(host),
+                trace: host.trace?.bind(host),
+                require: host.require?.bind(host),
+                // Sys
+                args: host.args,
+                newLine: host.newLine,
+                write: host.write.bind(host),
+                writeOutputIsTTY: host.writeOutputIsTTY?.bind(host),
+                getWidthOfTerminal: host.getWidthOfTerminal?.bind(host),
+                createHash: host.createHash?.bind(host),
+                createSHA256Hash: host.createSHA256Hash?.bind(host),
+                getMemoryUsage: host.getMemoryUsage?.bind(host),
+                exit: host.exit.bind(host),
+                enableCPUProfiler: host.enableCPUProfiler?.bind(host),
+                disableCPUProfiler: host.disableCPUProfiler?.bind(host),
+                cpuProfilingEnabled: host.cpuProfilingEnabled?.bind(host),
+                getEnvironmentVariable: host.getEnvironmentVariable.bind(host),
+                tryEnableSourceMapsForHost: host.tryEnableSourceMapsForHost?.bind(host),
+                debugMode: host.debugMode,
+                clearScreen: host.clearScreen?.bind(host),
+                setBlocking: host.setBlocking?.bind(host),
+                base64decode: host.base64decode?.bind(host),
+                base64encode: host.base64encode?.bind(host),
+                bufferFrom: host.bufferFrom?.bind(host),
+                defaultWatchFileKind: host.defaultWatchFileKind?.bind(host),
+                now: host.now?.bind(host),
+                disableUseFileVersionAsSignature: host.disableUseFileVersionAsSignature,
+                storeFilesChangingSignatureDuringEmit: host.storeFilesChangingSignatureDuringEmit,
+                // VirtualFS
+                ensureFileOrFolder: fshost.ensureFileOrFolder.bind(fshost),

```

Comment:
Instead of storing `ensureFileOrFolder` store private readonly fsHost on project service to be used in updateFileSystem only.